### DISCUSSION TITLE
What is the best practice for designing a workflow of dependent work orders

### DISCUSSION THREAD

**Discussion by hakancunier** [elsa-workflows/elsa-core#4538] [Q&A]
I want achive to process some serial work orders for producing a steel material/component.

First it should be cut, then it should be shaped, then it should be painted.

I have workorders like;
Id       | Operation  | StartAt | ParentId |
-------|-------------|---------|-----------|
100    | Cut             | 11:22   | null          |
123    | Shape        | 11:41   | 100          |
134    | Paint          | 13:00   | 123          |

So i draw a flow like;
Start > Operation Task (Cut) > Operation Task (Shape) > Operation Task (Paint) > End

Wrote custom activities for OperationTaskActivity.

Well at this point; there is no guide or i couldnt find some information about processing the Input or workflow variables.

Should i use WorkflowInput having a variable of first workOrder Id (100), if so what will be the input of OperationTask for shaping.
Or should i use contextId starting with 100, then updated within OperationTaskActivity(or workflowContextProvider) to 123 then 134.
Or should i use a variable currentWorkOrderId starting with 100 through 134.

Also i need to provide some information ie: processedQuantity from Cut through Paint. Should i use Variables for this?

Since I know my question seems specific to my use case, as i said above there is no guide to model this data flow. What would be the best approach for this use case.

Thank you.

**Comment by sfmskywalker** [comment]
If you need to process existing work orders, then I'm not sure if it makes sense to model a workflow after it, assuming that you will have different work orders in different order at different times. If so, I would probably just implement a order processor service that takes the first order, schedules a job, and when that job completes, move on to the next.

But, if you did want to create static workflows because maybe your work orders do not change every day and you need more control, then a workflow might make more sense.

In that case, I would probably provide both an **input** property called "Component" of type "Component (or whatever makes sense for your domain) as well as an **output** property that returns the modified component.

And if each operation can produce specific metadata about the operation that occurred, I would probably expose that as a separate output property.

This way, you have a uniform input/output interface for each activity representing an operation, making it easy to pass it through each step.

For Elsa 2, here's what this might look like (partially pseudo code):

```csharp
public abstract class Operation : Activity
{
   [ActivityInput] public Component Component { get; set; }
   [ActivityOutput] public Component Output { get;set; }
}

public sealed class Cut : Operation
{
   [ActivityOutput] public CutReport Report { get; set; }
}

public sealed class Shape : Operation
{
   [ActivityOutput] public ShapeReport Report { get; set; }
}

public sealed class Paint : Operation
{
   [ActivityOutput] public PaintReport Report { get; set; }
}
```

Now, when you add the activities to the workflow, you would probably provide the component to process via workflow input.
Then the first activity will use that input as its value for its Component property.
The second activity will then use the output of the first activity as its input.
The third activity will use the output of the second activity.

In Elsa 3, this is going to be even simpler where you don't even need to specify from where to get the input for a given activity; you would just need to specify that you want to use the output of the previous activity.
